contracts/OgvStaking.sol

```diff
@@ -108,54 +125,77 @@ contract OgvStaking is ERC20Votes {
+    /// @notice Unstake multiple lockups at once.
+    /// @param lockupIds Array of the lockup IDs to unstake
+    /// @return unstakedAmount OGV amount unstaked
+    /// @return rewardCollected OGV reward amount collected
+    function unstake(uint256[] memory lockupIds) external returns (uint256 unstakedAmount, uint256 rewardCollected) {
+        return _unstake(msg.sender, lockupIds);
     }
 
-    /// @notice Collect staked OGV for a lockup and any earned rewards.
-    /// @param lockupId the id of the lockup to unstake
-    function unstake(uint256 lockupId) external {
-        Lockup memory lockup = lockups[msg.sender][lockupId];
-        uint256 amount = lockup.amount;
-        uint256 end = lockup.end;
-        uint256 points = lockup.points;
-        require(block.timestamp >= end, "Staking: End of lockup not reached");
-        require(end != 0, "Staking: Already unstaked this lockup");
-        _collectRewards(msg.sender);
-        delete lockups[msg.sender][lockupId]; // Keeps empty in array, so indexes are stable
-        _burn(msg.sender, points);
-        ogv.transfer(msg.sender, amount);
-        emit Unstake(msg.sender, lockupId, amount, end, points);
+    /// @notice Unstakes lockups of an user.
+    ///         Can only be called by the Migrator.
+    /// @param staker Address of the user
+    /// @param lockupIds Array of the lockup IDs to unstake
+    /// @return unstakedAmount OGV amount unstaked
+    /// @return rewardCollected OGV reward amount collected
+    function unstakeFrom(address staker, uint256[] memory lockupIds)
+        external
+        onlyMigrator
+        returns (uint256 unstakedAmount, uint256 rewardCollected)
+    {
+        return _unstake(staker, lockupIds);
+    }
+
+    /// @notice Unstakes lockups of an user.
+    /// @param staker Address of the user
+    /// @param lockupIds Array of the lockup IDs to unstake
+    /// @return unstakedAmount OGV amount unstaked
+    /// @return rewardCollected OGV reward amount collected
+    function _unstake(address staker, uint256[] memory lockupIds)
+        internal
+        returns (uint256 unstakedAmount, uint256 rewardCollected)
+    {
+        if (lockupIds.length == 0) {
+            revert NoLockupsToUnstake();
+        }
+
+        // Collect rewards
+        rewardCollected = _collectRewards(staker);
+
+        for (uint256 i = 0; i < lockupIds.length; ++i) {
+            uint256 lockupId = lockupIds[i];
+            Lockup memory lockup = lockups[staker][lockupId];
+            uint256 amount = lockup.amount;
+            uint256 end = lockup.end;
+            uint256 points = lockup.points;
+
+            unstakedAmount += amount;
+
+            // Make sure it isn't unstaked already
+            if (end == 0) {
+                revert AlreadyUnstaked(lockupId);
+            }
+
+            delete lockups[staker][lockupId]; // Keeps empty in array, so indexes are stable
+
+            _burn(staker, points);
```

Comment:
Since lots of users have lots of stakes, let's total up the points to be burned and the OGV to be transferred, and do them all at the end. Will save a bunch of gas.

Reply:
Good point, will make that change

Reply:
Done